warehouse/query-core/src/main/java/datawave/query/planner/rules/NodeTransformVisitor.java

```diff
@@ -0,0 +1,125 @@
+import org.apache.commons.jexl2.parser.ASTNRNode;
+import org.apache.commons.jexl2.parser.ASTNotNode;
+import org.apache.commons.jexl2.parser.ASTOrNode;
+import org.apache.commons.jexl2.parser.ASTReference;
+import org.apache.commons.jexl2.parser.ASTReferenceExpression;
+import org.apache.commons.jexl2.parser.JexlNode;
+
+import java.util.List;
+
+public class NodeTransformVisitor extends RebuildingVisitor {
+    
+    private final ShardQueryConfiguration config;
+    private final List<NodeTransformRule> rules;
+    
+    public NodeTransformVisitor(ShardQueryConfiguration config, List<NodeTransformRule> rules) {
+        assert (rules != null);
+        assert (config != null);
+        this.config = config;
+        this.rules = rules;
+    }
+    
+    public static ASTJexlScript transform(ASTJexlScript tree, List<NodeTransformRule> rules, ShardQueryConfiguration config) {
+        NodeTransformVisitor visitor = new NodeTransformVisitor(config, rules);
+        return visitor.apply(tree);
+    }
+    
+    @Override
+    public Object visit(ASTOrNode node, Object data) {
+        return applyTransforms(super.visit(node, data));
+    }
+    
+    @Override
+    public Object visit(ASTAndNode node, Object data) {
+        return applyTransforms(super.visit(node, data));
+    }
+    
+    @Override
+    public Object visit(ASTEQNode node, Object data) {
+        return applyTransforms(super.visit(node, data));
+    }
+    
+    @Override
+    public Object visit(ASTNENode node, Object data) {
+        return applyTransforms(super.visit(node, data));
+    }
+    
+    @Override
+    public Object visit(ASTLTNode node, Object data) {
+        return applyTransforms(super.visit(node, data));
+    }
+    
+    @Override
+    public Object visit(ASTGTNode node, Object data) {
+        return applyTransforms(super.visit(node, data));
+    }
+    
+    @Override
+    public Object visit(ASTLENode node, Object data) {
+        return applyTransforms(super.visit(node, data));
+    }
+    
+    @Override
+    public Object visit(ASTGENode node, Object data) {
+        return applyTransforms(super.visit(node, data));
+    }
+    
+    @Override
+    public Object visit(ASTERNode node, Object data) {
+        return applyTransforms(super.visit(node, data));
+    }
+    
+    @Override
+    public Object visit(ASTNRNode node, Object data) {
+        return applyTransforms(super.visit(node, data));
+    }
+    
+    @Override
+    public Object visit(ASTNotNode node, Object data) {
+        return applyTransforms(super.visit(node, data));
+    }
+    
+    @Override
+    public Object visit(ASTFunctionNode node, Object data) {
+        return applyTransforms(super.visit(node, data));
+    }
+    
+    @Override
+    public Object visit(ASTReference node, Object data) {
+        // do not recurse on a marker node
+        if (QueryPropertyMarker.instanceOf(node, null)) {
+            return applyTransforms(node);
```

Comment:
Instead of applying the transform on the node that was passed in, we should instead apply the transforms to a copy of that node (using the base RebuildingVisitor), right?  That way we don't affect the original tree that was passed in.

Reply:
Also, should recursing on a marked node be configurable in case we someday have a transform which we expect to run against a marked node?

Reply:
Will do re: copy of node.  Interesting idea on the marked node.  I do not believe that is what is wanted in this case.  I think a marked node needs to always be handled as a single structure.  Applying a node transform outside the context of the marking is asking for trouble.